apps/devhub/tasks.py

```diff
@@ -32,65 +32,116 @@
+    """Run the validator against a file at the given path, and return the
+    results."""
+    if settings.VALIDATE_ADDONS:
+        upload = FileUpload.objects.get(pk=upload_id)
+        try:
+            result = run_validator(upload.path, listed=listed)
+        except Exception:
+            upload.update(task_error=traceback.format_exc())
+            raise
 
-    log.info('VALIDATING: %s' % upload_id)
-    upload = FileUpload.objects.using('default').get(pk=upload_id)
-    try:
-        result = run_validator(upload.path, listed=listed)
-    except Exception:  # Avoid bare except, which will catch system exceptions.
-        # Store the error with the FileUpload job, then raise
-        # it for normal logging.
-        tb = traceback.format_exception(*sys.exc_info())
-        upload.update(task_error=''.join(tb))
-        raise
+        # FIXME: Have validator return native Python datastructure.
+        return json.loads(result)
 
-    try:
-        # Try annotating the result with the "passed auto validation" status.
-        validation = json.loads(result)
-        # The automatic validation (which will automatically review and sign
-        # the file) is for unlisted (non-sideload) addons and beta versions.
-        # It only passes this automatic validation if there's no message with a
-        # signing_severity above trivial (so if there's any low/medium/high
-        # severity, it fails).
-        # Any "regular" messages from the amo-validator that should fail the
-        # auto validation now have a signing_severity level attached.
-        passed_auto_validation = False
-        if 'signing_summary' in validation:
-            summary = validation['signing_summary']
-            passed_auto_validation = (summary['low'] == 0 and
-                                      summary['medium'] == 0 and
-                                      summary['high'] == 0)
-        validation['passed_auto_validation'] = passed_auto_validation
-        result = json.dumps(validation)
-    except Exception:  # A bare `except:` would ignore system exceptions.
-        # Failed loading the result? No need to set passed_auto_validation.
-        pass
-
-    # Skip the "addon signed" warning if we're not signing.
-    if not settings.SIGNING_SERVER:
-        result = skip_signing_warning(result)
-    upload.validation = result
+
+@task(soft_time_limit=settings.VALIDATOR_TIMEOUT)
+def validate_file(file_id, **kw):
+    """Validate a File instance. If cached validation results exist, return
+    those, otherwise run the validator."""
+
+    if settings.VALIDATE_ADDONS:
+        file = File.objects.get(pk=file_id)
+        try:
+            return file.validation.validation
+        except FileValidation.DoesNotExist:
+            result = run_validator(file.current_file_path,
+                                   listed=file.version.addon.is_listed)
+            # FIXME: Have validator return native Python datastructure.
```

Comment:
I mean that the validator should just be returning the native datastructure that it generates rather than calling `json.dumps` and us calling `json.loads` on that.